File: src/main/java/vip/yazilim/spring/core/service/ACrudServiceImpl.java
package vip.yazilim.spring.core.service;

import java.util.List;
import java.util.Optional;

import org.springframework.data.jpa.repository.JpaRepository;

import vip.yazilim.spring.core.exception.general.InvalidArgumentException;
import vip.yazilim.spring.core.exception.general.InvalidUpdateException;
import vip.yazilim.spring.core.exception.general.database.DatabaseCreateException;
import vip.yazilim.spring.core.exception.general.database.DatabaseDeleteException;
import vip.yazilim.spring.core.exception.general.database.DatabaseException;
import vip.yazilim.spring.core.exception.general.database.DatabaseReadException;
import vip.yazilim.spring.core.exception.general.database.DatabaseUpdateException;

/**
 * Abstract Implementation of ICrudService.
 * <p>
 * Basic CRUD operations for an entity is implemented.
 *
 * @param <E> type entity
 * @param <ID> type of identity of entity
 * @author Emre Sen, 19.07.2019
 * @contact maemresen@yazilim.vip
 */
public abstract class ACrudServiceImpl<E, ID> implements ICrudService<E, ID> {

    /**
     * Repository of the entity that will be used with this implementation
     *
     * @return repository
     */
    protected abstract JpaRepository<E, ID> getRepository();

    /**
     * Get Id of the entity
     *
     * @param entity input entity model
     * @return id of entity
     */
    protected abstract ID getId(E entity);

    @Override
    public E save(E entity) throws DatabaseException {
        JpaRepository<E, ID> repository = getRepository();

        // save entity
        return repository.save(entity);
    }

    @Override
    public E create(E entity) throws DatabaseException {

        // initialize entity to insert
        // e.g setting unique UUID
        E initializedEntity = preInsert(entity);

        try {
            return save(initializedEntity);
        } catch (Exception e) {
            throw new DatabaseCreateException(e);
        }
    }

    /**
     * Operations before insert entity to the table
     * <p>
     * E.g.: setting uuid value
     *
     * @param entity values that will be inserted
     * @return entity that inserted to the data source
     */
    protected E preInsert(E entity) {
        return entity;
    }

    @Override
    public E update(E newEntity) throws DatabaseException, InvalidUpdateException, InvalidArgumentException {
        // get old entity
        ID id = getId(newEntity);
        Optional<E> oldEntity = getById(id);

        if (!oldEntity.isPresent()) {
            throw new InvalidUpdateException("Cannot update non-exist enity");
        }

        // prepare entity for update
        E preparedEntity = preUpdate(oldEntity.get(), newEntity);

        try {
            return save(preparedEntity);
        } catch (Exception e) {
            throw new DatabaseUpdateException(e);
        }
    }

    @Override
    public List<E> getAll() throws DatabaseException {
        JpaRepository<E, ID> repository = getRepository();
        try {
            return repository.findAll();
        } catch (Exception e) {
            throw new DatabaseReadException(e);
        }
    }

    @Override
    public Optional<E> getById(ID id) throws DatabaseException, InvalidArgumentException {
        if (id == null) {
            throw new InvalidArgumentException("ID cannot be null");
        }
        JpaRepository<E, ID> repository = getRepository();
        try {
            // find entity by id
            return repository.findById(id);
        } catch (Exception exception) {
            throw new DatabaseReadException(exception);
        }
    }

    /**
     * Operations before update operation
     *
     * @param oldEntity old data on data source
     * @param newEntity new data that will be saved to data source
     * @return prepared entity to update
     */
    protected E preUpdate(E oldEntity, E newEntity) {
        return newEntity;
    }

    @Override
    public boolean deleteById(ID id) throws DatabaseException {
        try {
            JpaRepository<E, ID> repository = getRepository();

            // delete entity
            repository.deleteById(id);
            return true;
        } catch (Exception exception) {
            throw new DatabaseDeleteException(exception);
        }
    }

    @Override
    public boolean delete(E entity) throws DatabaseException {
        try {
            JpaRepository<E, ID> repository = getRepository();

            // delete entity
            repository.delete(entity);
            return true;
        } catch (Exception exception) {
            throw new DatabaseDeleteException(exception);
        }
    }

    @Override
    public boolean deleteAll() throws DatabaseException {
        try {
            JpaRepository<E, ID> repository = getRepository();

            // delete entity
            repository.deleteAll();
            return true;
        } catch (Exception e) {
            throw new DatabaseDeleteException(e);
        }
    }
}

File: src/main/kotlin/vip/yazilim/libs/springvip/service/ACrudServiceImpl.kt
package vip.yazilim.libs.springvip.service

import org.jetbrains.annotations.NotNull
import org.springframework.data.jpa.repository.JpaRepository
import vip.yazilim.libs.springvip.exception.*
import java.util.*
import kotlin.reflect.KClass

/**
 * Abstract Implementation of ICrudService.
 *
 *
 * Basic CRUD operations for an entity is implemented.
 *
 * @param <E>  type entity
 * @param <ID> type of identity of entity
 * @author Emre Sen, 19.07.2019
 * @contact maemresen@yazilim.vip
</ID></E> */
abstract class ACrudServiceImpl<E : Any, ID : Any> : ICrudService<E, ID> {

    /**
     * Repository of the entity that will be used with this implementation
     * @return JPARepository Instance
     */
    protected abstract val repository: JpaRepository<E, ID>
    protected abstract val classOfEntity: KClass<E>

    /**
     * Get Id of the entity
     *
     * @param entity input entity model
     * @return id of entity
     */
    protected abstract fun getId( entity: E): ID

    @Throws(DatabaseSaveException::class)
    override fun save(entity: E): E {
        return try {
            repository.save(entity) ?: throw NoSuchElementException("Saved entity not found")
        } catch (exception: Exception) {
            throw DatabaseSaveException(entity::class, getId(entity), exception)
        }
    }

    @Throws(DatabaseCreateException::class)
    override fun create(entity: E): E {
        return try {
            // initialize entity to insert
            // e.g setting unique UUID
            val initializedEntity = preInsert(entity)
            save(initializedEntity)
        } catch (exception: Exception) {
            throw DatabaseCreateException(classOfEntity, exception)
        }
    }

    /**
     * Operations before insert entity to the table
     *
     *
     * E.g.: setting uuid value
     *
     * @param entity values that will be inserted
     * @return entity that inserted to the data source
     */
    protected open fun preInsert(entity: E): E {
        return entity
    }

    @Throws(DatabaseUpdateException::class)
    override fun update(newEntity: E): E {
        return try {
            // get old entity
            val id = getId(newEntity)
            val oldEntity = getById(id)
            require(oldEntity.isPresent) { "Cannot update non-exist entity" }

            // prepare entity for update
            val preparedEntity = preUpdate(oldEntity.get(), newEntity)
            save(preparedEntity)
        } catch (exception: Exception) {
            throw DatabaseUpdateException(classOfEntity, getId(newEntity), exception)
        }
    }

    override fun getAll(): List<E> {
        return try {
            // find entity by id
            repository.findAll()
        } catch (exception: Exception) {
            throw DatabaseReadException(classOfEntity, exception)
        }
    }

    @Throws(DatabaseReadException::class)
    override fun getById(id: ID): Optional<E> {
        return try {
            // find entity by id
            repository.findById(id)
        } catch (exception: Exception) {
            throw DatabaseReadException(classOfEntity, exception)
        }
    }

    /**
     * Operations before update operation
     *
     * @param oldEntity old data on data source
     * @param newEntity new data that will be saved to data source
     * @return prepared entity to update
     */
    protected open fun preUpdate(oldEntity: E, newEntity: E): E {
        return newEntity
    }

    @Throws(DatabaseDeleteException::class)
    override fun deleteById(id: ID): Boolean {
        return try {
            // delete entity
            repository.deleteById(id)
            true
        } catch (exception: Exception) {
            throw DatabaseDeleteException(classOfEntity, exception)
        }
    }

    @Throws(DatabaseDeleteException::class)
    override fun delete(entity: E): Boolean {
        return try {
            // delete entity
            repository.delete(entity)
            true
        } catch (exception: Exception) {
            throw DatabaseDeleteException(classOfEntity, exception)
        }
    }

    @Throws(DatabaseDeleteException::class)
    override fun deleteAll(): Boolean {
        return try {
            // delete entity
            repository.deleteAll()
            true
        } catch (exception: Exception) {
            throw DatabaseDeleteException(classOfEntity, exception)
        }
    }
}